assembl/static2/js/app/components/common/flatList.jsx

```diff
@@ -0,0 +1,125 @@
+  extractItems: Function
+};
+
+class FlatList extends React.Component<*, FlatListProps, void> {
+  props: FlatListProps;
+
+  loading: boolean;
+
+  static defaultProps = {
+    onEndReachedThreshold: 0.8,
+    itemData: () => {}
+  };
+
+  componentDidMount() {
+    window.addEventListener('scroll', this.handleScroll);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('scroll', this.handleScroll);
+  }
+
+  handleScroll = () => {
+    if (!this.loading) {
+      const { onEndReachedThreshold } = this.props;
+      // $FlowFixMe
+      const windowHeight = 'innerHeight' in window ? window.innerHeight : document.documentElement.offsetHeight;
+      const body = document.body;
+      const html = document.documentElement;
+      // $FlowFixMe
+      const docHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight);
+      const windowBottom = windowHeight + window.pageYOffset;
+      const threshold = docHeight * onEndReachedThreshold;
+      if (windowBottom >= threshold) {
+        this.onEndReached();
+      }
+    }
+  };
+
+  onEndReached = () => {
+    // The fetchMore method is used to load new data and add it
+    // to the original query we used to populate the list
+    const { fetchMore, items, extractItems, networkStatus } = this.props;
+    // If no request is in flight for this query, and no errors happened. Everything is OK.
+    if (networkStatus === APOLLO_NETWORK_STATUS.ready) {
+      this.loading = true;
+      fetchMore({
+        variables: { after: items.pageInfo.endCursor || '' },
+        updateQuery: (previousResult, { fetchMoreResult }) => {
+          // Don't do anything if there weren't any new items
+          const previousResultEntities = extractItems(previousResult);
+          if (!fetchMoreResult || !previousResultEntities.pageInfo.hasNextPage) {
+            return previousResult;
+          }
+          const fetchMoreResultEntities = extractItems(fetchMoreResult);
+          fetchMoreResultEntities.edges = previousResultEntities.edges.concat(fetchMoreResultEntities.edges);
+          return fetchMoreResult;
+        }
+      })
+        .then(() => {
+          this.loading = false;
+        })
+        .catch(() => {
+          displayAlert('danger', I18n.t('error.loading'));
+          this.loading = false;
+        });
+    }
+  };
+
+  onRefresh = () => {
+    const { refetch } = this.props;
+    refetch().catch(() => {
+      displayAlert('danger', I18n.t('error.loading'));
+    });
+  };
+
+  render() {
+    const { networkStatus, items, ListItem, itemData, className } = this.props;
+    if (
+      items == null ||
+      networkStatus === APOLLO_NETWORK_STATUS.loading ||
+      networkStatus === APOLLO_NETWORK_STATUS.setVariables
+    ) {
+      return <Loader color="black" />;
+    }
+    const entities = items.edges;
+    return (
+      <div className={className}>
+        {entities &&
+          entities.length > 0 &&
```

Comment:
why do you need `entities && entities.length > 0` ? I guess that `items.edges` may be `undefined` (or `null`?) but if not, it is an array and you don't need to check its length (`.map` will do nothing on an empty array)

Reply:
Yes, good catch... Done